matchmaking.py

```diff
@@ -0,0 +1,74 @@
+import random
+import time
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class Matchmaking:
+    def __init__(self, li, config, username):
+        self.li = li
+        self.variants = list(filter(lambda variant: variant != "fromPosition", config["challenge"]["variants"]))
+        self.matchmaking_cfg = config.get("matchmaking") or {}
+        self.username = username
+        self.last_challenge_created = time.time()
+        self.challenge_expire_time = 25  # The challenge expires 20 seconds after creating it.
+        self.challenge_id = None
+
+    def should_create_challenge(self):
+        matchmaking_enabled = self.matchmaking_cfg.get("allow_matchmaking")
+        time_has_passed = self.last_challenge_created + ((self.matchmaking_cfg.get("challenge_interval") or 30) * 60) < time.time()
+        challenge_expired = self.last_challenge_created + self.challenge_expire_time < time.time() and self.challenge_id
+        return matchmaking_enabled and (time_has_passed or challenge_expired)
+
+    def create_challenge(self, username, base_time, increment, days, variant):
+        mode = self.matchmaking_cfg.get("challenge_mode") or "random"
+        if mode == "random":
+            mode = random.choice(["casual", "rated"])
+        rated = mode == "rated"
+        params = {"rated": rated, "variant": variant}
+        if days:
+            params["days"] = days
+        else:
+            params["clock.limit"] = base_time
+            params["clock.increment"] = increment
+        challenge_id = self.li.challenge(username, params).get("challenge", {}).get("id")
+        return challenge_id
+
+    def choose_opponent(self):
+        variant = random.choice(self.variants)
+        base_time = self.matchmaking_cfg.get("challenge_initial_time", 60)
+        increment = self.matchmaking_cfg.get("challenge_increment", 2)
+        days = self.matchmaking_cfg.get("challenge_days")
+        game_duration = base_time + increment * 40
+        if variant != "standard":
+            game_type = variant
+        elif days:
+            game_type = "correspondence"
+        elif game_duration < 179:
+            game_type = "bullet"
+        elif game_duration < 479:
+            game_type = "blitz"
+        elif game_duration < 1499:
+            game_type = "rapid"
+        else:
+            game_type = "classical"
+
+        min_rating = self.matchmaking_cfg.get("opponent_min_rating") or 2000

```

Comment:
If a user doesn't want to filter based on rating, they should be able to put 0 as a minimum. So, this should be `.get("opponent_min_rating", 2000)` (would zero be a better default?). Even though making this change for the maximum rating would filter out every possible bot, perhaps the change should also be made there so these settings are treated the same.

Reply:
On lichess I believe the minimum rating is 600 and max is 4000 so those seem like reasonable default values. Maybe it could be mentioned that 600 is min and 4000 is max

Reply:
I changed the default values, so there is no need to change the line to `.get("opponent_min_rating", 2000)`. Now bots will challenge other bots without taking into account their rating, if the options are left to their default values.